runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java

```diff
@@ -180,14 +186,14 @@ public TimerInternals timerInternals() {
     DoFnRunnerWithMetrics<InputT, OutputT> doFnRunnerWithMetrics =
         new DoFnRunnerWithMetrics<>(stepName, doFnRunner, metricsAccum);
 
-    return new SparkProcessContext<>(
+    SparkProcessContext<Object, InputT, OutputT> ctx =
+        new SparkProcessContext<>(
```

Comment:
Just throwing it out here as option to consider ... considering there's already a lot of moving pieces involved and the context has just become a container without functionality, everything in there except for the runner could already be passed to the processor when initializing it (some for the input iterator as well). The only thing that then has to be passed to `process` is the runner itself.

Reply:
It actually can not be put into constructor, because processor is responsible for providing correct instance of output manager which is needed doFnRunner construction. Therefore I did choose to wrap these into container and pass to processor as context object.

Reply:
>  because processor is responsible for providing correct instance of output manager which is needed doFnRunner construction

That's more or less what I meant, you can but everything into the constructor of the processor except the runner itself. E.g. process could look like this then:
```
processor.process(iter, doFnRunnerWithMetrics)
```
or even as below if you pass the input iterator into the constructor as well.
```
processor.process(doFnRunnerWithMetrics)
```

Reply:
@JozoVilcek Please feel free to discard or ignore! [This code here](https://github.com/JozoVilcek/beam/commit/e055041b73331cd2d1c3161033906f49b9c361e8) is quickly hacked together and just meant to demonstrate what I had in mind above. `SparkInputDataProcessor` was replaced by `SparkOutputManager`, no `SparkProcessContext` and output iterator needed anymore.